In [2]:
%pip install pandas pyarrow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 13.8 MB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 17.2 MB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 347.8/347.8 KB 15.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.8/16.8 MB 15.8 MB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 509.2/509.2 KB 21.5 MB/s eta 0:00:00
You should consider upgrading via the '/home/hachan/.pyenv/versions/3.10.4/bin/python3.10 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [41]:
import pandas as pd
import numpy as np

## Q1. Downloading the data

We'll use [the same NYC taxi dataset](https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page),
but instead of "**Green** Taxi Trip Records", we'll use "**Yellow** Taxi Trip Records".

Download the data for January and February 2023.

Read the data for January. How many columns are there?

* 16
* 17
* 18
* 19

In [70]:
df = pd.read_parquet("data/yellow_tripdata_2023-01.parquet")

In [3]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00

In [4]:
df.describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  3.066766e+06                     3066766                     3066766   
mean   1.730215e+00  2023-01-17 00:22:26.288164  2023-01-17 00:38:06.427874   
min    1.000000e+00         2008-12-31 23:01:42         2009-01-01 14:29:11   
25%    1.000000e+00  2023-01-09 16:21:57.250000         2023-01-09 16:37:06   
50%    2.000000e+00  2023-01-17 08:42:29.500000  2023-01-17 08:58:30.500000   
75%    2.000000e+00         2023-01-24 16:26:27         2023-01-24 16:42:49   
max    2.000000e+00         2023-02-01 00:56:53         2023-02-02 09:28:47   
std    4.438480e-01                         NaN                         NaN   

       passenger_count  trip_distance    RatecodeID  PULocationID  \
count     2.995023e+06   3.066766e+06  2.995023e+06  3.066766e+06   
mean      1.362532e+00   3.847342e+00  1.497440e+00  1.663980e+02   
min       0.000000e+00   0.000000e+00  1.000000e+00  1.000000e+00   
25%       1.000000e+00   1.060000e+00  1.000000e+00  1.320000e+02   
50%       1.000000e+00   1.800000e+00  1.000000e+00  1.620000e+02   
75%       1.000000e+00   3.330000e+00  1.000000e+00  2.340000e+02   
max       9.000000e+00   2.589281e+05  9.900000e+01  2.650000e+02   
std       8.961200e-01   2.495838e+02  6.474767e+00  6.424413e+01   

       DOLocationID  payment_type   fare_amount         extra       mta_tax  \
count  3.066766e+06  3.066766e+06  3.066766e+06  3.066766e+06  3.066766e+06   
mean   1.643926e+02  1.194483e+00  1.836707e+01  1.537842e+00  4.882900e-01   
min    1.000000e+00  0.000000e+00 -9.000000e+02 -7.500000e+00 -5.000000e-01   
25%    1.140000e+02  1.000000e+00  8.600000e+00  0.000000e+00  5.000000e-01   
50%    1.620000e+02  1.000000e+00  1.280000e+01  1.000000e+00  5.000000e-01   
75%    2.340000e+02  1.000000e+00  2.050000e+01  2.500000e+00  5.000000e-01   
max    2.650000e+02  4.000000e+00  1.160100e+03  1.250000e+01  5.316000e+01   
std    6.994368e+01  5.294762e-01  1.780782e+01  1.789592e+00  1.034641e-01   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
count  3.066766e+06  3.066766e+06           3.066766e+06  3.066766e+06   
mean   3.367941e+00  5.184907e-01           9.820847e-01  2.702038e+01   
min   -9.622000e+01 -6.500000e+01          -1.000000e+00 -7.510000e+02   
25%    1.000000e+00  0.000000e+00           1.000000e+00  1.540000e+01   
50%    2.720000e+00  0.000000e+00           1.000000e+00  2.016000e+01   
75%    4.200000e+00  0.000000e+00           1.000000e+00  2.870000e+01   
max    3.808000e+02  1.969900e+02           1.000000e+00  1.169400e+03   
std    3.826759e+00  2.017579e+00           1.833529e-01  2.216359e+01   

       congestion_surcharge   airport_fee  
count          2.995023e+06  2.995023e+06  
mean           2.274231e+00  1.074086e-01  
min           -2.500000e+00 -1.250000e+00  
25%            2.500000e+00  0.000000e+00  
50%            2.500000e+00  0.000000e+00  
75%            2.500000e+00  0.000000e+00  
max            2.500000e+00  1.250000e+00  
std            7.718454e-01  3.556511e-01

In [5]:
df.shape[1]

19

In [63]:
df.shape[0]

3066766

# Question 2 Computing duration
Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

- 32.59
- 42.59
- 52.59
- 62.59


In [72]:
df['duration'] = (df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60

In [11]:
df['duration'].std()

np.float64(42.59435124195458)

## Q3. Dropping outliers

Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

* 90%
* 92%
* 95%
* 98%

In [12]:
df_norm = df[df['duration'].between(1, 60)]

In [15]:
df_norm.shape[0] / df.shape[0]

0.9812202822125979

## Q4. One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model. 

* Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will 
  label encode them)
* Fit a dictionary vectorizer 
* Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

* 2
* 155
* 345
* 515
* 715

In [18]:
x_dict = df_norm[['PULocationID', 'DOLocationID']].astype(str).to_dict(orient='records')

In [23]:
from sklearn.feature_extraction import DictVectorizer

In [25]:
dv = DictVectorizer()
m = dv.fit_transform(x_dict)

In [26]:
m

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 6018346 stored elements and shape (3009173, 515)>

## Q5. Training a model

Now let's use the feature matrix from the previous step to train a model. 

* Train a plain linear regression model with default parameters, where duration is the response variable
* Calculate the RMSE of the model on the training data

What's the RMSE on train?

* 3.64
* 7.64
* 11.64
* 16.64

In [38]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [33]:
model = LinearRegression()

In [36]:
y_train = df_norm['duration']

In [37]:
model.fit(m, y_train)

LinearRegression()

In [39]:
y_pred = model.predict(m)

In [60]:
rmse1 = np.sqrt(mean_squared_error(y_train, y_pred))
rmse1

np.float64(7.649261822035489)

## Q6. Evaluating the model

Now let's apply this model to the validation dataset (February 2023). 

What's the RMSE on validation?

* 3.81
* 7.81
* 11.81
* 16.81


In [71]:
dfval = pd.read_parquet("data/yellow_tripdata_2023-02.parquet")

In [62]:
dfval.shape[0]

2913955

In [51]:
dfval['duration'] = (dfval['tpep_dropoff_datetime'] - dfval['tpep_pickup_datetime']).dt.total_seconds() / 60

In [64]:
dfval_norm = dfval[dfval['duration'].between(1, 60)]

In [65]:
xval_dict = dfval_norm[['PULocationID', 'DOLocationID']].astype(str).to_dict(orient='records')

In [66]:
m_val = dv.transform(xval_dict)

In [67]:
yval = dfval_norm['duration']

In [68]:
yval_pred = model.predict(m_val)

In [69]:
rmse2 = np.sqrt(mean_squared_error(yval, yval_pred))
rmse2

np.float64(7.811821332387183)